### PR TITLE
[XPU] fix correctness issue of GDN triton kernel for XPU

### PR DESCRIPTION
**Pull request by Xia-Weiwen**
## Motivation

fix correctness issue of GDN triton kernel for XPU for long sequence length.

## Modifications

Kernels in python/sglang/srt/hardware_backend/xpu/kernels/fla/chunk_delta_h.py and python/sglang/srt/hardware_backend/xpu/kernels/fla/chunk_fwd.py.
Test cases in test/registered/attention/test_chunk_gated_delta_rule.py

## Accuracy Tests

qwen3.5-4B on Intel B60, gsm8k, 100 questions
Accuracy: 0.810
Invalid: 0.000

qwen3.5-9B on Intel B60, gsm8k, 1319 questions
Accuracy: 0.895 (SOTA=0.904)
Invalid: 0.001 (SOTA=0.001)

qwen3.5-35B-A3B on Intel B60, gsm8k, 1319 questions
Accuracy: 0.926
Invalid: 0.000

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :hourglass_flowing_sand: [Run #26381025046](https://github.com/sgl-project/sglang/actions/runs/26381025046)
Latest PR Test (Extra): :x: [Run #26381025011](https://github.com/sgl-project/sglang/actions/runs/26381025011)

### CODE REVIEW

**Comment by Xia-Weiwen** [comment]
Hi @mingfeima Please review. Thanks.

File: python/sglang/srt/hardware_backend/xpu/kernels/fla/chunk_fwd.py [hunk 1]
```diff
@@ -285,6 +295,8 @@ def chunk_gated_delta_rule_fwd_intra(
     NT = triton.cdiv(T, BT) if cu_seqlens is None else len(chunk_indices)
 
     # Step 1: fused kkt + solve_tril
+    # Use float32 for A to avoid precision loss in intermediate scratch storage.
+    # Cast back to input dtype before passing to recompute_w_u_fwd.
     A = torch.zeros(B, T, H, BT, device=k.device, dtype=k.dtype)
```
[the remark this answers is not in the document]
Comment (Xia-Weiwen) [reply python/sglang/srt/hardware_backend/xpu/kernels/fla/chunk_fwd.py:298]: Fixed

File: python/sglang/srt/hardware_backend/xpu/kernels/fla/chunk_fwd.py [hunk 2]
```diff
@@ -302,6 +314,10 @@ def chunk_gated_delta_rule_fwd_intra(
         BC=BC,
     )
 
+    # Cast A back to input dtype: recompute_w_u_fwd does tl.dot(b_A, b_vb)
+    # which requires matching dtypes.
+    # A = A.to(k.dtype)
```
[the remark this answers is not in the document]
Comment (Xia-Weiwen) [reply python/sglang/srt/hardware_backend/xpu/kernels/fla/chunk_fwd.py]: Fixed

File: python/sglang/srt/hardware_backend/xpu/kernels/fla/chunk_delta_h.py [hunk 3]
```diff
@@ -110,67 +113,108 @@ def chunk_gated_delta_rule_fwd_kernel_h_blockdim64_k_loop(
...
-            )
-            tl.store(p_ht, b_h1.to(p_ht.dtype.element_ty), boundary_check=(0, 1))
+            # Save updated h to scratch (initial_state) for next time step
+            if INPLACE_UPDATE:
+                p_hs = tl.make_block_ptr(
+                    ht, (V, K), (K, 1), (i_v * BV, k_blk), (BV, 64), (1, 0)
+                )
+                tl.store(p_hs, b_h.to(p_hs.dtype.element_ty), boundary_check=(0, 1))
+            else:
+                assert False, "currently we always do in-place update"
```
[the remark this answers is not in the document]
Comment (Xia-Weiwen) [reply python/sglang/srt/hardware_backend/xpu/kernels/fla/chunk_delta_h.py]: Fixed

File: test/registered/attention/test_chunk_gated_delta_rule.py [hunk 4]
```diff
@@ -210,6 +210,21 @@ def test_multi_chunk_nt8(self):
...
+
+    def test_long_prompt_1536(self):
+        self._check_shape(B=1, T_per_seq=1536, H=16, K=128, V=128, pool_size=32)
+
+    def test_long_prompt_2048(self):
+        self._check_shape(B=1, T_per_seq=2048, H=16, K=128, V=128, pool_size=32)
+
+    def test_long_prompt_batch(self):
+        self._check_shape(B=2, T_per_seq=1024, H=16, K=128, V=128, pool_size=32)
+
```
Comment (mingfeima) [inline test/registered/attention/test_chunk_gated_delta_rule.py:222]: nit: `pytest.mark.parametrize`

**Review by mingfeima — APPROVED** [review]

File: test/registered/attention/test_chunk_gated_delta_rule.py (hunk 4 above)
Comment (Xia-Weiwen) [reply test/registered/attention/test_chunk_gated_delta_rule.py:222]: Simplified with a for loop
